crypto3/libs/multiprecision/include/nil/crypto3/multiprecision/big_uint_impl.hpp

```diff
@@ -0,0 +1,1674 @@
+    constexpr auto operator&(const T1& a, const T2& b) noexcept {
+        largest_t result = a;
+        largest_t::bitwise_and(result, b);
+        return result;
+    }
+    NIL_CO3_MP_BIG_UINT_INTEGRAL_ASSIGNMENT_TEMPLATE
+    constexpr auto& operator&=(big_uint_t& a, const T& b) {
+        big_uint_t::bitwise_and(a, b);
+        return a;
+    }
+
+    NIL_CO3_MP_BIG_UINT_INTEGRAL_TEMPLATE
+    constexpr auto operator|(const T1& a, const T2& b) noexcept {
+        largest_t result = a;
+        largest_t::bitwise_or(result, b);
+        return result;
+    }
+    NIL_CO3_MP_BIG_UINT_INTEGRAL_ASSIGNMENT_TEMPLATE
+    constexpr auto& operator|=(big_uint_t& a, const T& b) {
+        big_uint_t::bitwise_or(a, b);
+        return a;
+    }
+
+    NIL_CO3_MP_BIG_UINT_INTEGRAL_TEMPLATE
+    constexpr auto operator^(const T1& a, const T2& b) noexcept {
+        largest_t result = a;
+        largest_t::bitwise_xor(result, b);
+        return result;
+    }
+    NIL_CO3_MP_BIG_UINT_INTEGRAL_ASSIGNMENT_TEMPLATE
+    constexpr auto& operator^=(big_uint_t& a, const T& b) {
+        big_uint_t::bitwise_or(a, b);
+        return a;
+    }
+
+    NIL_CO3_MP_BIG_UINT_UNARY_TEMPLATE
+    constexpr auto operator~(const big_uint_t& a) noexcept {
+        big_uint_t result;
+        big_uint_t::complement(result, a);
+        return result;
+    }
+
+    NIL_CO3_MP_BIG_UINT_UNARY_TEMPLATE
+    constexpr auto operator<<(const big_uint_t& a, std::size_t shift) noexcept {
+        big_uint_t result = a;
+        big_uint_t::left_shift(result, shift);
+        return result;
+    }
+    NIL_CO3_MP_BIG_UINT_UNARY_TEMPLATE
+    constexpr auto& operator<<=(big_uint_t& a, std::size_t shift) noexcept {
+        big_uint_t::left_shift(a, shift);
+        return a;
+    }
+
+    NIL_CO3_MP_BIG_UINT_UNARY_TEMPLATE
+    constexpr auto operator>>(const big_uint_t& a, std::size_t shift) noexcept {
+        big_uint_t result = a;
+        big_uint_t::right_shift(result, shift);
+        return result;
+    }
+    NIL_CO3_MP_BIG_UINT_UNARY_TEMPLATE
+    constexpr auto& operator>>=(big_uint_t& a, std::size_t shift) noexcept {
+        big_uint_t::right_shift(a, shift);
+        return a;
+    }
+
+#undef NIL_CO3_MP_BIG_UINT_UNARY_TEMPLATE
+#undef NIL_CO3_MP_BIG_UINT_INTEGRAL_ASSIGNMENT_TEMPLATE
+#undef NIL_CO3_MP_BIG_UINT_INTEGRAL_TEMPLATE
+
+    // Hash
+
+    template<std::size_t Bits>
+    constexpr std::size_t hash_value(const big_uint<Bits>& val) noexcept {
+        std::size_t result = 0;
+        for (std::size_t i = 0; i < val.limbs_count(); ++i) {
+            boost::hash_combine(result, val.limbs()[i]);
+        }
+        return result;
+    }
+
+    // IO
+
+    template<std::size_t Bits>
+    std::ostream& operator<<(std::ostream& os, const big_uint<Bits>& value) {
+        os << value.str(os.flags());
+        return os;
+    }
+
+    // Common ops
+
+    template<std::size_t Bits>
+    constexpr std::size_t msb(const big_uint<Bits>& a) {
```

Comment:
Why do we need this helper? Maybe just change all the calls of msb(a) to a.msb() and delete this, and other helpers?

Reply:
Yeah. So there is some code that should work for both big_uint and built-in integers. This code uses these free functions.